planner/implementation/simple_plans.go

```diff
@@ -140,6 +140,16 @@ func (impl *TiDBTopNImpl) CalcCost(outCount float64, children ...memo.Implementa
 	return impl.cost
 }
 
+// AttachChildren implements Implementation AttachChildren interface.
+func (impl *TiDBTopNImpl) AttachChildren(children ...memo.Implementation) memo.Implementation {
+	topN := impl.plan.(*plannercore.PhysicalTopN)
+	topN.SetChildren(children[0].GetPlan())
+	// If the topN.ByItems contains ScalarFunctions, we need to inject extra Projections
+	// below and above TopN for them.
+	impl.plan = plannercore.InjectProjBelowSort(topN, topN.ByItems)
```

Comment:
If you put it here, how we eliminate the two projection in tests?

Reply:
It cannot be eliminated by the current optimizer either. If we make this as a transformation rule, it might be more complex:
1. if we erase the old TopN after injecting Projection, the TopN maybe can not be be pushed down Gather. (unless TiKV supports Projection).
2. if we do not erase the old TopN, this rule can be applied many times when the TopN can be pushed down to other Operands.